src/MediaUploader.php

```diff
@@ -438,12 +453,51 @@ public function upload()
 
         $this->verifyDestination($model);
 
-        $this->filesystem->disk($model->disk)->put($model->getDiskPath(), $this->source->contents());
+        $source = $this->callBefore('before_upload', $model, SourceAdapterInterface::class);
+
+        if ($source !== null) {
+            $model->size = $source->size();
+            $model->mime_type = $source->mimeType();
+            $model->extension = $source->extension();
+            $model->aggregate_type = $this->inferAggregateType($model->mime_type, $model->extension);
+        } else {
+            $source = $this->source;
+        }
+
+        $alteredModel = $this->callBefore('before_save', $model, Model::class);
```

Comment:
I wouldn't make this callback return anything, instead, just let the user modify the model record passed to the callback. Objects are passed by reference so the changes will persist.

Reply:
True, it's done, but I don't know how to test it know since there's nothing that can be changed without being kind of idiotoc in the default Media model.

Reply:
you could test manually setting the ID on the model and make sure that is what you get back.